go/vt/schema/online_ddl.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2019 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+
+	"vitess.io/vitess/go/vt/topo"
+)
+
+var (
+	migrationBasePath = "schema-migration"
+)
+
+// MigrationBasePath is the root for all schema migration entries
+func MigrationBasePath() string {
+	return migrationBasePath
+}
+
+// MigrationRequestsPath is the base path for all newly received schema migration requests.
+// such requests need to be investigates/reviewed, and to be assigned to all shards
+func MigrationRequestsPath() string {
+	return fmt.Sprintf("%s/requests", MigrationBasePath())
+}
+
+// MigrationReviewedPath is the base path for schema migrations that have been reviewed. Kept for
+// historical reference
+func MigrationReviewedPath() string {
+	return fmt.Sprintf("%s/reviewed", MigrationBasePath())
+}
+
+// MigrationJobsKeyspacePath is the base path for a tablet job, by keyspace
+func MigrationJobsKeyspacePath(keyspace string) string {
+	return fmt.Sprintf("%s/jobs/%s", MigrationBasePath(), keyspace)
+}
+
+// MigrationJobsKeyspaceShardPath is the base path for a tablet job, by keyspace and shard
+func MigrationJobsKeyspaceShardPath(keyspace, shard string) string {
+	return fmt.Sprintf("%s/%s", MigrationJobsKeyspacePath(keyspace), shard)
+}
+
+// OnlineDDLStatus is an indicator to a online DDL status
+type OnlineDDLStatus string
+
+const (
+	OnlineDDLStatusRequested OnlineDDLStatus = "requested"
+	OnlineDDLStatusReviewed  OnlineDDLStatus = "reviewed"
+	OnlineDDLStatusCancelled OnlineDDLStatus = "cancelled"
+	OnlineDDLStatusQueued    OnlineDDLStatus = "queued"
+	OnlineDDLStatusReady     OnlineDDLStatus = "ready"
+	OnlineDDLStatusRunning   OnlineDDLStatus = "running"
+	OnlineDDLStatusComplete  OnlineDDLStatus = "complete"
+	OnlineDDLStatusFailed    OnlineDDLStatus = "failed"
+)
+
+// OnlineDDL encapsulates the relevant information in an online schema change request
+type OnlineDDL struct {
+	Keyspace    string          `json:"keyspace,omitempty"`
+	Table       string          `json:"table,omitempty"`
+	SQL         string          `json:"sql,omitempty"`
+	UUID        string          `json:"uuid,omitempty"`
```

Comment:
with current design, each keyspace/shard will be handed this migration. All shards will use same UUID as the "migration job id". Perhaps I should rename the variable to `JobID` or something.
This will make it easier to investigate issues; if you have the UUID and you know all tables/shards use that same UUID, then all logs are in same place, called by same names, etc. on all tablets.

One key design is that a tablet/master will not run two concurrent migrations. Now, that's subtle. Running two long running migrations is known to be slower than running the two sequentially. However, sometimes one will be running a 3 day long migration, and then also want to `ALTER` a very small table, which only takes 2 minutes to run. That's a valid use case to running concurrently. Right now I'm not looking into that, and enforce serialization. If `vitess` is opinionated, and recommends a max `250-300GB` of data per shard, then migration time is also capped to a reasonable timeframe (a few hours) and we can afford to serialize everything.